File: Leetcode/combinationsum.java
import java.util.*;

public class combinationsum {
    
}class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        Arrays.sort(candidates);

        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates, result, new ArrayList<>(), target, 0);
        return result;
    
        
    }

    public void backtrack(int[] candidates, List<List<Integer>> result, List<Integer> cur, int target, int start){
        if(target ==0){
            result.add(new ArrayList<>(cur));
        }
        if(target<0){
            return;
        }

        for(int i=start;i<candidates.length;i++){
            cur.add(candidates[i]);
            backtrack(candidates, result, cur, target-candidates[i], i);
            cur.remove(cur.size()-1);

        }
    }
}

File: LCmac/combinationsum.py


from ast import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        res = []
        def dfs(i, cur, total):
            if total == target:
                res.append(cur.copy())
                return
            if i >= len(candidates) or total > target:
                return 
            
            # if we choose to add candidates[i]
            cur.append(candidates[i])

            dfs(i, cur, total + candidates[i])

            # if we choose not to add candidates[i]

            cur.pop()

            dfs(i+1, cur, total)

        dfs(0, [], 0)

        return res
                
        